src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -594,15 +602,21 @@ private CheckSumAndSizeWriteStream GetDataCompressor(Stream backingStream, bool
         {
             // stream stack: backingStream -> DeflateStream -> CheckSumWriteStream
 
-            // we should always be compressing with deflate. Stored is used for empty files, but we don't actually
-            // call through this function for that - we just write the stored value in the header
-            Debug.Assert(CompressionMethod == CompressionMethodValues.Deflate);
-
-            Stream compressorStream = _compressionLevel.HasValue ?
-                new DeflateStream(backingStream, _compressionLevel.Value, leaveBackingStreamOpen) :
-                new DeflateStream(backingStream, CompressionMode.Compress, leaveBackingStreamOpen);
-
             bool isIntermediateStream = true;
+            Stream compressorStream;
+            if (!_compressionLevel.HasValue)
```

Comment:
I would first set the value to Compress if it has no value. Then test the value, using a switch statement.